railties/test/engine/commands_test.rb

```diff
@@ -64,17 +64,20 @@ def test_server_command_broadcast_logs
       pid = spawn_command("server", replica, env: { "RAILS_ENV" => "development" })
       assert_output("Listening on", primary)
 
-      Net::HTTP.new("127.0.0.1", 3000).tap do |net|
-        net.get("/")
+      thread = Thread.new do
+        Net::HTTP.new("127.0.0.1", 3000).tap do |net|
+          net.get("/")
+        end
```

Comment:
Oh by the way, why was this not simply?
```rb
Net::HTTP.get("127.0.0.1", "/", 3000)
```

Reply:
I'm not sure but @Edouard-chin added it in #49417, so maybe he has more context on this particular test. I'm surprised he did't run into this, but maybe was not using a mac at the time?

Reply:
Thank you for the great investigation, TIL about pty and the limitation on various platform ! Sorry I introduced the bug 😭 , I didn't realise as I didn't encounter it, strangely, even now on a mac. 


> Oh by the way, why was this not simply? [...]

No particular reason, feel free to change it!

----------------

For a suggestion: while your fix is great, I realise now that I don't think this test needs PTY (and the rest of this test suit neither I believe), it just needs a pair of connected IO. I tried a few dozen times, and it works correctly, but please try on your machine

```diff
diff --git a/railties/test/engine/commands_test.rb b/railties/test/engine/commands_test.rb
index 3e888ce0d8..97c7a6d1f0 100644
--- a/railties/test/engine/commands_test.rb
+++ b/railties/test/engine/commands_test.rb
@@ -60,9 +60,10 @@ def test_server_command_work_inside_engine
     end

     def test_server_command_broadcast_logs
-      primary, replica = PTY.open
-      pid = spawn_command("server", replica, env: { "RAILS_ENV" => "development" })
-      assert_output("Listening on", primary)
+      read, write = IO.pipe
+      pid = spawn_command("server", write, env: { "RAILS_ENV" => "development" })
+
+      assert_output("Listening on", read)

       Net::HTTP.new("127.0.0.1", 3000).tap do |net|
         net.get("/")
@@ -73,7 +74,7 @@ def test_server_command_broadcast_logs
         assert_match("Processing by Rails::WelcomeController", logs)
       end

-      assert_output("Processing by Rails::WelcomeController", primary)
+      assert_output("Processing by Rails::WelcomeController", read)
     ensure
       kill(pid)
     end

```

Reply:
@Edouard-chin Thanks for your comment!

> I didn't encounter it, strangely, even now on a mac.

That is indeed strange! Have you tried running [this `ptsbufsize.py` script](https://superuser.com/a/1452858) locally?

<details><summary>Here's a Ruby version too if that's easier to run.</summary>

```rb
#!/usr/bin/env ruby

require "pty"
require "timeout"

primary, replica = PTY.open

puts "Using devices #{primary.path} ; #{replica.path} as PTY."

begin
  bytes_written = 0
  loop do
    Timeout.timeout(1) do
      replica.write("a")
    end
    bytes_written += 1
  end
rescue Timeout::Error
  puts "Timed out after writing #{bytes_written} bytes."
end
```

</details> 

On my machine it stops and outputs this:
```
Using devices masterpty:/dev/ttys010 ; /dev/ttys010 as PTY.
Timed out after writing 1024 bytes.
```

I tested in Terminal.app, iTerm2 and Alacritty, in and out of tmux, and always got the same result.
(Supposedly, those don't make a difference since `PTY` would be using macOS' TTY/PTY devices...)

Reply:
@Edouard-chin Thanks again for the suggestion!

I applied the changes you suggested and it looks like the test still works. That also means we don't need to check `available_pty?` for this specific test.

After looking at all tests in the file, it looks like only the `dbconsole` one actually needs to run in a pseudo-terminal.
All the others seem to be doing just fine with `IO.pipe`, so I changed them to do so.

Reply:
Thanks for the script, I can reproduce with it 100% of the time with it.
I also wrote a script to run the test in a loop and got it to hang after the test ran 4 times. I suppose there is a race condition which makes the test succeed if the buffer reads the "Listening on" (and thus add available space to the buffer) before the get request is hit.